### ISSUE TITLE
Quotes inside interpolations can terminate the interpolated string

### ISSUE THREAD

**Issue by panicbit** [nushell/nushell#12722]
### Describe the bug

When interpolating something into a string, patterns like `'"'` (single quote, double quote, single quote) can terminate the interpolated string.

### How to reproduce

1. `$"('"')"`

In interactive mode, nushell asks for more input.
In non-interactive mode nushell errors out with:

```
Error: nu::parser::unexpected_eof

  × Unexpected end of code.
   ╭─[/tmp/test.nu:2:1]
 1 │ $"('"')"
   ╰────
```

Adding a `'` to the end (`$"('"')"'`) results in the output `""'`.

### Expected behavior

Output: `"`

### Configuration

| key                | value                                                                           |
| ------------------ | ------------------------------------------------------------------------------- |
| version            | 0.93.0                                                                          |
| major              | 0                                                                               |
| minor              | 93                                                                              |
| patch              | 0                                                                               |
| branch             |                                                                                 |
| commit_hash        |                                                                                 |
| build_os           | linux-x86_64                                                                    |
| build_target       | x86_64-unknown-linux-gnu                                                        |
| rust_version       | rustc 1.77.2 (25ef9e3d8 2024-04-09)                                             |
| rust_channel       | stable-x86_64-unknown-linux-gnu                                                 |
| cargo_version      | cargo 1.77.2 (e52e36006 2024-03-26)                                             |
| build_time         | 2024-05-01 16:35:43 +02:00                                                      |
| build_rust_channel | release                                                                         |
| allocator          | mimalloc                                                                        |
| features           | dataframe, default, sqlite, static-link-openssl, system-clipboard, trash, which |
| installed_plugins  | formats                                                                         |

**Comment by devyn** [comment]
Yep, that doesn't seem right. Will have to look at the parse tree.

**Comment by ItsHarper** [comment]
I'm encountering a similar issue that is probably related, but not quite identical. Putting it here because it's probably best fixed at the same time as the original one.

Simplified reproduction command: 
```nushell
print $"(if (random dice --sides 2 | get 0) == 1 { "got heads" } else { "got tails" })"
```

Strangely, it only breaks **if the internal strings have spaces**. This version works fine (only one of the strings needs to have a space to break it):
```nushell
print $"(if (random dice --sides 2 | get 0) == 1 { "heads" } else { "tails" })"
```

**Comment by isti115** [comment]
The examples given for the issue with handling whitespace (I've tried with space, tab and newline, they all cause failure) can be further simplified into just the following:

Works:
```nu
> $"("_")"
# _
```

Fails:
```nu
> $"(" ")"
# Error: nu::parser::parse_mismatch
# 
#   × Parse mismatch during operation.
#    ╭─[entry #22:1:6]
#  1 │ $"(" ")"
#    ·      ─┬─
#    ·       ╰── expected operator
#    ╰────
```
